validator/client/beacon-api/activation.go

```diff
@@ -0,0 +1,166 @@
+//go:build use_beacon_api
+// +build use_beacon_api
+
+package beacon_api
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"context"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/api/gateway/apimiddleware"
+	rpcapi "github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+func (c beaconApiValidatorClient) waitForActivation(ctx context.Context, in *ethpb.ValidatorActivationRequest) (ethpb.BeaconNodeValidator_WaitForActivationClient, error) {
+	return &beaconNodeValidatorWaitForActivationClient{
+		ctx:                        ctx,
+		beaconApiValidatorClient:   c,
+		ValidatorActivationRequest: in,
+	}, nil
+}
+
+type beaconNodeValidatorWaitForActivationClient struct {
+	grpc.ClientStream
+	ctx context.Context
+	beaconApiValidatorClient
+	*ethpb.ValidatorActivationRequest
+	lastRecvTime time.Time
+}
+
+func (x *beaconNodeValidatorWaitForActivationClient) Recv() (*ethpb.ValidatorActivationResponse, error) {
+	now := time.Now()
+
+	nextRecvTime := x.lastRecvTime.Add(time.Duration(params.BeaconConfig().SecondsPerSlot) * time.Second)
+	if x.lastRecvTime.IsZero() {
+		nextRecvTime = now
+	}
+
+	waitDuration := nextRecvTime.Sub(now)
+
+	select {
+	case <-time.After(waitDuration):
+		x.lastRecvTime = nextRecvTime
+
+		// Represents the target set of keys
+		targetPubKeys := make(map[string]struct{}, len(x.ValidatorActivationRequest.PublicKeys))
+
+		// Represents the set of keys actually returned by the beacon node
+		retrievedPubKeys := make(map[string]struct{})
+
+		// Contains all keys in targetPubKeys but not in retrievedPubKeys
+		missingPubKeys := make(map[string]struct{})
+
+		statuses := []*ethpb.ValidatorActivationResponse_Status{}
+		params := url.Values{}
+
+		for _, publicKey := range x.ValidatorActivationRequest.PublicKeys {
+			stringPubKey := hexutil.Encode(publicKey)
+			targetPubKeys[stringPubKey] = struct{}{}
+			params.Add("id", stringPubKey)
+		}
+
+		resp, err := x.beaconApiValidatorClient.httpClient.Get(x.beaconApiValidatorClient.url + "/eth/v1/beacon/states/head/validators?" + params.Encode())
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if err = resp.Body.Close(); err != nil {
+				return
+			}
+		}()
+
+		if resp.StatusCode != http.StatusOK {
```

Comment:
Http request and responses should be abstracted into their own method and away from the main `WaitForActivation` loop

Reply:
Created dedicated function for request/response `getStateValidators`.